### PR TITLE
Moving the shared enums to gfx_traits; r=jdm

### PR DESCRIPTION
**Pull request by ghost**
This is just a draft for showing my initial commit for [issue #5698](https://github.com/servo/servo/issues/5698). I'll continue updating the commit until the build succeeds :) /cc @jdm

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/6685)

### CODE REVIEW

File: components/gfx_traits/paint_task.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! The task that handles all painting.
+
+use azure::azure_hl::Color;
+use display_list::StackingContext;
```
Comment (jdm) [inline components/gfx_traits/paint_task.rs]: Why is this here?

File: components/gfx/display_list/mod.rs [hunk 2]
```diff
@@ -17,6 +17,7 @@
 #![deny(unsafe_code)]
 
 use display_list::optimizer::DisplayListOptimizer;
+use gfx_traits::display_list::*;
```
Comment (jdm) [inline components/gfx/display_list/mod.rs]: Why is this change necessary?

File: components/gfx/paint_task.rs [hunk 3]
```diff
@@ -39,60 +39,34 @@ use util::task::spawn_named_with_send_on_failure;
...
-}
-
-pub struct PaintRequest {
-    pub buffer_requests: Vec<BufferRequest>,
-    pub scale: f32,
-    pub layer_id: LayerId,
-    pub epoch: Epoch,
-    pub layer_kind: LayerKind,
-}
+use gfx_traits::paint_task::{CompositorPaintMsg, PaintTask};
```
Comment (jdm) [inline components/gfx/paint_task.rs]: PaintTask doesn't exist yet, right?

File: components/gfx/platform/freetype/font_template.rs [hunk 4]
```diff
@@ -5,16 +5,7 @@
...
-/// Platform specific font representation for Linux.
-/// The identifier is an absolute path, and the bytes
-/// field is the loaded data that can be passed to
-/// freetype and azure directly.
-#[derive(Deserialize, Serialize)]
-pub struct FontTemplateData {
-    pub bytes: Vec<u8>,
-    pub identifier: Atom,
-}
+use gfx_traits::platform::font_template::freetype::FontTemplateData;
```
Comment (jdm) [inline components/gfx/platform/freetype/font_template.rs]: Why this change?

File: components/gfx/platform/macos/font_template.rs [hunk 5]
```diff
@@ -13,24 +13,7 @@ use std::borrow::ToOwned;
...
-    ///
-    /// When sending a `FontTemplateData` instance across processes, this will be set to `None` on
-    /// the other side, because `CTFont` instances cannot be sent across processes. This is
-    /// harmless, however, because it can always be recreated.
-    pub ctfont: CachedCTFont,
-
-    pub identifier: Atom,
-    pub font_data: Option<Vec<u8>>
-}
+use gfx_traits::platform::font_template::macos::FontTemplateData;
```
Comment (jdm) [inline components/gfx/platform/macos/font_template.rs]: Why this change?

File: components/gfx/text/glyph.rs [hunk 6]
```diff
@@ -3,6 +3,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 use euclid::point::Point2D;
+use gfx_traits::text::glyph::*;
```
Comment (jdm) [inline components/gfx/text/glyph.rs]: Why this change?

File: components/gfx/text/text_run.rs [hunk 7]
```diff
@@ -4,6 +4,7 @@
 
 use font::{Font, FontHandleMethods, FontMetrics, IS_WHITESPACE_SHAPING_FLAG, RunMetrics};
 use font::{ShapingOptions};
+use gfx_traits::text::text_run::*;
```
Comment (jdm) [inline components/gfx/text/text_run.rs]: Why this change?

**Comment by jdm** [comment]
This is a good start, but I think there's a lot of code being moved around here this isn't necessary, and I suspect it stems from the unnecessary inclusion of `StackingContext` in `gfx_traits/paint_task.rs`.

**Comment by Wafflespeanut** [comment]
@jdm Initially, I had moved the `StackingContext` inside `gfx_traits` because `Msg` demanded it. Now that we've split the `Msg` enum, it isn't of any use anymore (and, I forgot about that). So yeah, you're totally right. Will make this work today :)

**Comment by Wafflespeanut** [comment]
@jdm r? (and, I shall proceed to PaintTask and font_cache_task) :)

**Comment by jdm** [comment]
Much better!

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 3 of 20 files at r1, 20 of 20 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[components/gfx/paint_task.rs, line 262 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6685#-JupmHL85nI3v0WgYPAt)** ([raw file](https://github.com/servo/servo/blob/49d0e11b21f5437d2947fc691aaf1102875d03fd/components/gfx/paint_task.rs#L262)):</sup>
Move this to a new line.

---

<sup>**[components/gfx_traits/lib.rs, line 4 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6685#-Jupn3UjInjOlYNmbU4_)** ([raw file](https://github.com/servo/servo/blob/5caf1301692c5b53c300ef13a83a8f2e8c1519bd/components/gfx_traits/lib.rs#L4)):</sup>
Whoops!

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6685)

**Comment by Wafflespeanut** [comment]
@jdm Missed that. Fixed now :)

**Comment by jdm** [comment]
@bors-servo: r+
-S-awaiting-review +S-awaiting-merge   

---

Reviewed 2 of 2 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6685)

**Comment by jdm** [comment]
Missing appropriate changes to the various Cargo.lock files.

**Comment by Wafflespeanut** [comment]
@jdm Build succeeds locally. r?

**Comment by jdm** [comment]
You can also use the power of rebase: http://rypress.com/tutorials/git/rebasing

**Comment by Wafflespeanut** [comment]
@jdm Ooooh, nice! Thanks. (sadly, /me has already done the exhaustive merging -_-)

**Comment by Wafflespeanut** [comment]
@jdm Ooooh, nice! Thanks. (sadly, /me has already done the exhaustive merging -_-). Anyways, r?

**Comment by pcwalton** [comment]
Unfortunately, I think this should wait on e10s. It'll be a lot of churn otherwise.

**Comment by Wafflespeanut** [comment]
@pcwalton Oh, take your time. Ping me when you're done :)

**Comment by metajack** [comment]
@pcwalton Which PR should this be blocked on? Or did you mean all e10s PRs?

**Comment by Wafflespeanut** [comment]
Looks like this PR hasn't got much work anymore (since @pcwalton has done most of it). All there's left is to just move the public `ChromeToPaintMsg` enum to `gfx_traits`.

**Comment by Wafflespeanut** [comment]
Alright, this doesn't work yet! `msg` depends on `canvas_traits` which depends on `gfx_traits` which further depends on `msg`. So, cyclic dependency!

**Comment by Wafflespeanut** [comment]
Alright, this doesn't work yet! There's a cyclic dependency! Here's how they depend on one another:

- `msg::constellation_msg` --> `canvas_traits::CanvasMsg`
- `canvas_traits` --> `gfx_traits::color`
- `gfx_traits::paint_task` --> `msg::{constellation_msg, compositor_msg}`

And, none of the enums can be moved from one to another, because they appear as a *core* enum in all these crates (i.e., they're being used by a lot of other modules in other crates).

So, I can't proceed @jdm :/

**Comment by jdm** [comment]
The only solution I've come up with is splitting the giant constellation_msg::Msg enum into FromScript and FromCompositor enums, and moving the FromScript one to `script_traits`. This would mean adding a second receiver to the constellation, and making it select over both of them, etc., but it would remove the `msg -> canvas_traits` dependency. Fun times!

**Comment by Wafflespeanut** [comment]
@jdm Once we split the `Msg` and add a new receiver to `Constellation`, it'd then require a sender, for which we'll need to create something (say 'ScriptConstellationChan') similar to the `ConstellationChan` wrapper struct (for utilizing `MsgFromScript`). Now, both the wrapper structs should be added to `Constellation`.

But, the [`start`](https://mxr.mozilla.org/servo/source/components/compositing/constellation.rs#231) method returns a `ConstellationChan`, which doesn't include the `ScriptConstellationChan`. How should I overcome this problem?

And, it'd make a considerable difference in my view if I understand [this `impl`](https://mxr.mozilla.org/servo/source/components/compositing/constellation.rs#221). I can't even see where this `start` method is used :/

**Comment by jdm** [comment]
http://mxr.mozilla.org/servo/source/components/servo/lib.rs#169 is where Constellation::start gets used.

**Comment by jdm** [comment]
I'm imagining the new channel being stored inside Constellation and being passed to new pipelines that are created.

**Comment by Wafflespeanut** [comment]
@jdm Awesome! Thanks :)

**Comment by Wafflespeanut** [comment]
@jdm Thanks. Now, this gets interesting - [canvasrenderingcontext2d.rs](https://mxr.mozilla.org/servo/source/components/script/dom/canvasrenderingcontext2d.rs#125) and [webglrenderingcontext.rs](https://mxr.mozilla.org/servo/source/components/script/dom/webglrenderingcontext.rs#71) are the only files that are gonna be using the variants from `ScriptConstellationChan`, but they seem to get the thing from `GlobalRef`, which means both the `Worker` and the `Window` must have another field for `ScriptConstellationChan`, right? So, this is gonna get messy.

Once that's done, I could go for a `global.constellation_chan(args)` where `args` would typically be a `bool` asking whether to supply a `CompositorConstellationChan` or `ScriptConstellationChan` or I could also go for the good old procedure of adding another method for `script_constellation_chan` to the global. Either way, this is gonna get complicated, as it's probably gonna modify a great deal of files. Any ideas?

**Comment by jdm** [comment]
@Wafflespeanut I'm pretty sure that Worker and Window could replace their constellation chan with the new ScriptConstellationChan, if the name actually describes its purpose :)

**Comment by Wafflespeanut** [comment]
Meh...

**Comment by Wafflespeanut** [comment]
@jdm Now, there's another iceberg. `ScriptTask` creates a `NewLayoutTaskInfo` using its own [`constellation_chan`](https://mxr.mozilla.org/servo/source/components/script/script_task.rs#1092) (which is now a `ScriptConstellationChan`). But, it can't have that, because it's then used to create a [`LayoutTask`](https://mxr.mozilla.org/servo/source/components/layout/layout_task.rs#646), which impl's from a `LayoutTaskFactory` trait, whose `create` method relies on a `CompositorConstellationChan`.

I'm not sure whether I'm supposed to change the entire trait, and change those layout task related stuff. Any suggestions?

(oh, and sorry to smack you with this on your return from vacation :P)

**Comment by jdm** [comment]
I suspect layout could move to using a ScriptConstellationChan (which might be more aptly named ContentConstellationChan, in that case)?

**Comment by jdm** [comment]
So yeah, I would try modifying the trait to use ScriptConstellationChan, which will require moving the StylesheetLoadComplete, ReflowComplete, and TickAllAnimations messages from ConstellationControlMsg into MsgFromScript.

**Comment by Wafflespeanut** [comment]
(sigh), forgot the recently added tidy checks

**Comment by jdm** [comment]
As I feared, we have a lot of code moving around here and we still can't achieve my original goal of making `compositing` depend entirely on `gfx_traits` and not `gfx`. This is because of the `LayoutToPaint` msg that is in `gfx` that `compositing` relies upon.

**Comment by Wafflespeanut** [comment]
@jdm Yep, you were right. But, isn't there any way (at all?) to break that dependency? Only the `LayoutToPaintMsg` is stopping us.

**Comment by Wafflespeanut** [comment]
@jdm I guess, we can't move only the `Msg` after all.

`LayoutToPaintMsg` (which is now in gfx_traits) demands `StackingContext`, whose impl requires PaintContext (which forces us to move the whole paint_context.rs) and then, PaintContext needs filters and `FontContext` (so, we gotta move font_context.rs) - now, `FontContext` needs font, font_cache_task, font_template & platform (and some of them demand text too).

Does this mean we'll be moving everything other than the remains of paint_task to gfx_traits? Am I complicating something, or is this the way it's always been? Because, I fear whether the future gfx crate would only have paint_task :)

**Comment by jdm** [comment]
I haven't come up with any great solutions. Sorry! Thanks for settling the question of whether this was actually feasible, though!

**Comment by Wafflespeanut** [comment]
@jdm No problem, I had fun time with it :)
